### PR TITLE
Add a hash map for storing groupBy partition index

### PR DESCRIPTION
**Pull request by drcrallen**
* Improves groupBy performance by approx 15%

### CODE REVIEW

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java [hunk 1]
```diff
@@ -148,7 +148,8 @@ public void close() throws IOException
     private final BufferAggregator[] aggregators;
     private final PositionMaintainer positionMaintainer;
 
-    private final TreeMap<ByteBuffer, Integer> positions;
+    private final Map<ByteBuffer, Integer> positions = Maps.newTreeMap();
```
Comment (xvrl) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: is there a reason to keep the treeMap at all? If ordering is required, would a linkedhashmap be sufficient?

Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: Looking at the code, it seems like a treemap isn't required and we can use a hashmap for positions and avoid the positionsHash variable

Comment (fjy) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: I believe the sorted ordering is important for iteration

**Comment by fjy** [comment]
+1 I believe this is merge-able

**Comment by xvrl** [comment]
Before we merge can we add comments to indicate the purpose of the second hash map. It wouldn't be obvious out of context.

File: processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java (hunk 1 above)
Comment (drcrallen) [inline processing/src/main/java/io/druid/query/groupby/GroupByQueryEngine.java]: It iterates into a linked hash map later. I'm not completely clear what the bytebuffer is in this key case. I'm assuming it's the bytes comparable for the timestamp, but I'm not 100% certain. I find it odd that the treemap is ordering on the bytebuffer without having some sort of custom comparator.

**Comment by drcrallen** [comment]
@xvrl : comment added

**Pull request merged**